pubsub/google/cloud/pubsub_v1/publisher/_batch/thread.py

```diff
@@ -137,6 +137,17 @@ def status(self):
         """
         return self._status
 
+    def wait(self):
+        """If commit in progress, waits until all of the futures resolved.
+
+        .. note::
+
+            This method blocks until all futures of this batch resolved.
+        """
+        if self._status in (base.BatchStatus.STARTING, base.BatchStatus.IN_PROGRESS):
```

Comment:
I think we have to worry about races with this conditional, but not with the for-loop itself.

1) There is a race between a batch going from ACCEPTING_MESSAGES to STARTING in commit() and this check for the state being STARTING.
2) If a batch is in the state SUCCESS or ERROR, this for-loop wont run. So, wait() may return before the futures return with a result.

It would be good to take the lock to prevent these races.

How about:

should_wait = false
with self._state_lock:
            if self._status != base.BatchStatus.ACCEPTING_MESSAGES:
                    should_wait = true
if should_wait:
        for future in self._futures:
                future.result()

I would also worry about races in modifications to self._futures. However, in the way this class is set up, reads of self._futures only take place after all writes are done (aka, the class is no longer in state ACCEPTING_MESSAGES). The code I gave above maintains this invariant. This is not obvious. I wish we could document it.

What do you think?